package-res/resources/web/pentaho/visual/base/model.js

```diff
@@ -45,37 +73,287 @@ define([
+            doExecute = result.value.doExecute;
+            dataFilter = result.value.dataFilter;
+          }
+
+          if(doExecute) {
+            try {
+              doExecute(dataFilter);
+              result = new ActionResult(undefined, null);
+            } catch(e) {
+              result = new ActionResult(undefined, e);
+            }
+          } else {
+            result = new ActionResult(undefined, new UserError(bundle.structured.error.action.notDefined));
+          }
+          return this._broadcastResult(DidExecute, RejectedExecute, result, dataFilter);
+
+        },
+
+        /**
+         * Emits an event. All exceptions are caught (and swallowed).
+         * @param {!pentaho.lang.Event} event - The event to be emitted.
+         * @returns {pentaho.lang.Event} If successful returns `event`, otherwise returns `null`.
+         *
+         * @override
+         * @protected
+         */
+        _emit: function(event) {
+          var result = null;
+          try {
+            result = this.base(event);
+          } catch(e) {
+            console.log("Exception thrown during '", event.type, "' loop:", e);
+          }
+          return result;
+        },
+
+        /**
+         * Processes a "will:" event and returns a [result]{@link pentaho.lang.ActionResult}
+         * that wraps a possibly mutated event.
+         *
+         * If no listener is registered for the "will:" event, returns `null`.
+         * If some event listener canceled the event (and thus the action at play),
+         * the result is an [user error]{pentaho.lang.UserError} that captures the reason for the cancelation.
+         *
+         * This method abstracts the first part of the "will/did/rejected"
+         * set of events associated with an action.
+         *
+         * @param {pentaho.visual.base.Will} WillEvent - The constructor of the "will:*" event to be processed.
+         * @param RejectedEvent - The constructor of the "rejected:" event associated with `WillEvent`.
+         * @param {pentaho.data.filter.AbstractFilter} dataFilter - The filter describing the data being processed.
+         * @param {*} otherArg - The third argument to be passed to the constructor of `WillEvent`.
+         * @return {pentaho.lang.ActionResult}
+         * If no listener for the "will:" event was registered, `null` is returned.
+         * If unsuccessful, the `error` field contains the exception that originated the error.
+         * If successful, the `error` field is `null` and the `value` field contains the `WillEvent` instance.
+         *
+         * @protected
+         *
+         * @see #_broadcastResult
+         */
+        _runPipeline: function(WillEvent, RejectedEvent, dataFilter, otherArg) {
+          if(!this._hasListeners(WillEvent.type)) return null;
+
+          var will = new WillEvent(this, dataFilter, otherArg);
+          var isCanceled = !this._emit(will);
+
+          if(isCanceled) {
```

Comment:
I agree with the redundancy on the handling of the canceled event, but I think the _runPipeline method has a reason to be here. Perhaps the better approach is to do something else. Given the fact that there three event loops invoking the _runPipeline/_broadcastResult methods, we should at least try to avoid the repeated code. Perhaps a single method implementing the whole event loop but accepting a callback for the part in the middle?